libmobilecoin/src/transaction.rs

```diff
@@ -1139,6 +1192,352 @@ pub extern "C" fn mc_memo_sender_with_payment_request_memo_get_payment_request_i
+/// to 58 valid UTF-8 bytes. The actual note stored on chain is up to 57 null
+/// terminated UTF-8 bytes unless the note is exactly 57 utf-8 bytes long,
+/// in which case, no null bytes are stored. If the C string passed here is
+/// exactly 58 bytes, the last byte MUST be null and that byte will be
+/// removed prior to storage on chain.
+#[no_mangle]
+pub extern "C" fn mc_memo_builder_gift_code_sender_create(
+    gift_code_sender_note: FfiStr,
+) -> FfiOptOwnedPtr<McTxOutMemoBuilder> {
+    ffi_boundary(|| {
+        let note = <&str>::try_from_ffi(gift_code_sender_note)
+            .expect("Failed to decode gift_code_sender note");
+        let gift_code_sender_memo_builder = GiftCodeSenderMemoBuilder::new(note)
+            .expect("Gift code sender note had invalid utf-8 or was more than 57 bytes long");
+
+        let memo_builder_box: Box<dyn MemoBuilder + Sync + Send> =
+            Box::new(gift_code_sender_memo_builder);
+
+        Some(memo_builder_box)
+    })
+}
+
+/// # Preconditions
+///
+/// * `global_index` - must be the global TxOut index of the originally funded
+///   gift code TxOut
+#[no_mangle]
+pub extern "C" fn mc_memo_builder_gift_code_cancellation_create(
+    global_index: u64,
+) -> FfiOptOwnedPtr<McTxOutMemoBuilder> {
+    ffi_boundary(|| {
+        let gift_code_cancellation_memo_builder =
+            GiftCodeCancellationMemoBuilder::new(global_index);
+
+        let memo_builder_box: Box<dyn MemoBuilder + Sync + Send> =
+            Box::new(gift_code_cancellation_memo_builder);
+
+        Some(memo_builder_box)
+    })
+}
+
+/* ==== GiftCodeFundingMemo ==== */
+
+/// # Preconditions
+///
+/// * `tx_out_public_key` - must be a valid 32-byte Ristretto-format scalar.
+/// * `fee` - must be an integer less than or equal to 2^56
+/// * `gift_code_funding_note` - must be a null-terminated C string containing
+/// up to 54 valid UTF-8 bytes. The actual note stored on chain is up to 53 null
+/// terminated UTF-8 bytes unless the note is exactly 53 utf-8 bytes long,
+/// in which case, no null bytes are stored. If the C string passed here is
+/// exactly 54 bytes, the last byte MUST be null and that byte will be
+/// removed prior to storage on chain.
+/// * `out_memo_data` - length must be >= 64.
+///
+/// # Errors
+///
+/// * `LibMcError::InvalidInput`
+#[no_mangle]
+pub extern "C" fn mc_memo_gift_code_funding_memo_create(
+    tx_out_public_key: FfiRefPtr<McBuffer>,
+    fee: u64,
+    gift_code_funding_note: FfiStr,
+    out_memo_data: FfiMutPtr<McMutableBuffer>,
+    out_error: FfiOptMutPtr<FfiOptOwnedPtr<McError>>,
+) -> bool {
+    ffi_boundary_with_error(out_error, || {
+        let note = <&str>::try_from_ffi(gift_code_funding_note)
+            .expect("Failed to decode gift code funding note");
+        let key = RistrettoPublic::try_from_ffi(&tx_out_public_key)?;
+        let memo = GiftCodeFundingMemo::new(&key, fee, note).unwrap();
```

Comment:
Here it makes sense to keep the unwrap because the ffi will log the error thrown by the new method, and the errors that can be thrown by this method are more diverse than just note length.